packages/cli/src/utils/checkpoint.ts

```diff
@@ -0,0 +1,177 @@
+ * if an error occurs it will silently fail and return undefined values
+ */
+export async function tryToReadDataFromSchema(schemaPath: string) {
+  let schemaProvider: string | undefined
+  let schemaPreviewFeatures: string[] | undefined
+  let schemaGeneratorsProviders: string[] | undefined
+
+  try {
+    const schema = await getSchema(schemaPath)
+    const config = await getConfig({
+      datamodel: schema,
+      ignoreEnvVarErrors: true,
+    })
+
+    if (config.datasources.length > 0) {
+      schemaProvider = config.datasources[0].provider
+    }
+
+    // Example 'prisma-client-js'
+    schemaGeneratorsProviders = config.generators
+      // Check that value is defined
+      .filter((generator) => generator && generator.provider)
+      .map((generator) => parseEnvValue(generator.provider))
+
+    // restrict the search to previewFeatures of `provider = 'prisma-client-js'`
+    // (this was not scoped to `prisma-client-js` before Prisma 3.0)
+    const prismaClientJSGenerator = config.generators.find(
+      (generator) => parseEnvValue(generator.provider) === 'prisma-client-js',
+    )
+    if (prismaClientJSGenerator && prismaClientJSGenerator.previewFeatures.length > 0) {
+      schemaPreviewFeatures = prismaClientJSGenerator.previewFeatures
+    }
+  } catch (e) {
+    debug('Error from tryToReadDataFromSchema()')
+    debug(e)
+  }
+
+  return {
+    schemaProvider,
+    schemaPreviewFeatures,
+    schemaGeneratorsProviders,
+  }
+}
+
+/*
+ * String options of the CLI
+ * Tip: search for `: String,`
+ */
+export const SENSITIVE_CLI_OPTIONS = [
+  // 1. Connection strings
+  '--url',
+  '--shadow-database-url',
+  '--from-url',
+  '--to-url',
+  // 2. Paths
+  '--schema',
+  '--file',
+  '--from-schema-datamodel',
+  '--to-schema-datamodel',
+  '--from-schema-datasource',
+  '--to-schema-datasource',
+  '--from-migrations',
+  '--to-migrations',
+  '--hostname',
+  // 3. Migration names
+  '--name',
+  '--applied',
+  '--rolled-back',
+]
+/*
+ * removes potentially sensitive information from the command array (argv strings)
+ */
+export const redactCommandArray = (commandArray: string[]) => {
+  for (let i = 0; i < commandArray.length; i++) {
+    const arg = commandArray[i]
+    // redact --option arguments
+    SENSITIVE_CLI_OPTIONS.forEach((option: string) => {
+      // --url file:./dev.db
+      // arg is `--url` and a complete match
+      const argIndexCompleteMatch = arg === option
+      // --url=file:./dev.db
+      // arg value is `--url=file:./dev.db` and a partial match
+      const argIndexPartialMatch = arg.indexOf(option)
+
+      // First check for complete match and redact the value
+      if (argIndexCompleteMatch) {
+        commandArray[i + 1] = '[redacted]'
+      }
+      // else check for partial match and redact the value
+      else if (argIndexPartialMatch !== -1) {
+        commandArray[i] = `${option}=[redacted]`
+      }
+    })
+  }
+
+  return commandArray
+}
```

Comment:
This seems to be working, but it might be hard to read/extend this code and understanding the flow.
I'd suggest normalizing the commands first (getting rid of "=", which is a syntactic sugar that is discarded without any additional semantics meaning), and I'd avoid mutating the input array directly.

Consider something like:

```ts
export const redactCommandArray = (commandArray: string[]): string[] => {
  const SENSITIVE_CLI_OPTIONS_SET = new Set(SENSITIVE_CLI_OPTIONS)

  // ['cmd', '--applied', 'secret', '--url=secret'] -> ['cmd', '--applied', 'secret', '--url', 'secret']
  const normalizedCommandArray = commandArray.flatMap((cmd) => cmd.split('='))

  const indexesToRedact = normalizedCommandArray
    .map((cmd, i) => [SENSITIVE_CLI_OPTIONS_SET.has(cmd), i] as const)
    .filter(([isCmdSensitive, _]) => isCmdSensitive)
    .map(([_, i]) => i + 1)

  const REDACTED_TAG = '[redacted]'

  // replace items at `indexesToRedact` with `REDACTED_TAG`
  const redactedCommandArray = [...normalizedCommandArray]
  indexesToRedact.forEach(i => {
    redactedCommandArray[i] = REDACTED_TAG
  })

  return redactedCommandArray
}
```

Should you accept this change, the test case 'should redact --option=[value]' would become essentially identical to 'should redact --option [value]', with a `.toEqual(['cmd', option, '[redacted]'])`.
```suggestion
export const redactCommandArray = (commandArray: string[]) => {
  for (let i = 0; i < commandArray.length; i++) {
    const arg = commandArray[i]
    // redact --option arguments
    SENSITIVE_CLI_OPTIONS.forEach((option: string) => {
      // --url file:./dev.db
      // arg is `--url` and a complete match
      const argIndexCompleteMatch = arg === option
      // --url=file:./dev.db
      // arg value is `--url=file:./dev.db` and a partial match
      const argIndexPartialMatch = arg.indexOf(option)

      // First check for complete match and redact the value
      if (argIndexCompleteMatch) {
        commandArray[i + 1] = '[redacted]'
      }
      // else check for partial match and redact the value
      else if (argIndexPartialMatch !== -1) {
        commandArray[i] = `${option}=[redacted]`
      }
    })
  }

  return commandArray
}
```

Reply:
Since the connection strings likely have `=` chars this would mess up the `commandArray` by removing them.
So I don't think we can apply this as is, though I like it 👍🏼 

Examples
```
init --url="postgresql://janedoe:mypassword@localhost:5432/mydb?schema=sample"
init --url="sqlserver://localhost:1433;initial catalog=sample;user=sa;password=mypassword;"
```
https://www.prisma.io/docs/reference/database-reference/connection-urls

Reply:
Thank you for pointing out this edge case!